crates/extensions/tedge_config_manager/src/plugin.rs

```diff
@@ -0,0 +1,153 @@
+use crate::error::ConfigManagementError;
+use camino::Utf8Path;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+use std::process::Output;
+use std::sync::Arc;
+use tedge_api::CommandLog;
+use tedge_api::LoggedCommand;
+use tedge_config::SudoCommandBuilder;
+
+pub const LIST: &str = "list";
+const GET: &str = "get";
+const SET: &str = "set";
+
+#[derive(Debug, Clone)]
+pub struct ExternalPlugin {
+    pub name: String,
+    pub path: PathBuf,
+    pub sudo: SudoCommandBuilder,
+    tmp_dir: Arc<Utf8Path>,
+}
+
+impl ExternalPlugin {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        name: impl Into<String>,
+        path: impl Into<PathBuf>,
+        sudo: SudoCommandBuilder,
+        tmp_dir: Arc<Utf8Path>,
+    ) -> ExternalPlugin {
+        ExternalPlugin {
+            name: name.into(),
+            path: path.into(),
+            sudo,
+            tmp_dir,
+        }
+    }
+
+    pub fn command(&self, action: &str) -> Result<LoggedCommand, ConfigManagementError> {
+        let mut command = self.sudo.command(&self.path);
+        command.arg(action);
+
+        let command = LoggedCommand::from_command(command, self.tmp_dir.as_ref());
+
+        Ok(command)
+    }
+
+    pub async fn execute(
+        &self,
+        command: LoggedCommand,
+        command_log: Option<&mut CommandLog>,
+    ) -> Result<Output, ConfigManagementError> {
+        let output = command
+            .execute(command_log)
+            .await
+            .map_err(|err| self.plugin_error(err))?;
+
+        if !output.status.success() {
+            let stderr = String::from_utf8_lossy(&output.stderr);
+            return Err(ConfigManagementError::PluginError {
+                plugin_name: self.name.clone(),
+                reason: format!("Set command failed: {}", stderr),
```

Comment:
I think this `execute()` is called by not only `set`, but also `list` and `get`, so the error message is misleading?